File: tests/fixtures.py
"""Synthetic feed documents.

Written by hand rather than captured from Kagi, for two reasons: the repo stays
free of third-party article text, and a fixture can carry the cases real feeds
happen not to contain today -- a hostile href, a script tag, an item with no
picture, the same headline in two sections.
"""

from __future__ import annotations

import struct
import zlib


def channel(title: str, built: str, items: str) -> bytes:
    return (
        """<?xml version='1.0' encoding='UTF-8'?>
<rss xmlns:atom="http://www.w3.org/2005/Atom" version="2.0">
  <channel>
    <title>Kagi News - %s</title>
    <link>https://kite.kagi.com/x.xml</link>
    <lastBuildDate>%s</lastBuildDate>
    %s
  </channel>
</rss>"""
        % (title, built, items)
    ).encode()


def item(
    title: str,
    *,
    section: str = "World",
    sub: str = "Wildfires",
    pub: str = "Sun, 09 Aug 2026 17:13:45 +0000",
    body: str = "<p>First paragraph.</p><p>Second paragraph.</p>",
    image: str | None = "https://img.example/one.png",
    extras: str = "",
) -> str:
    img = (
        "&lt;img src='%s' alt='A caption for the picture.' /&gt;&lt;br /&gt;" % image
        if image
        else ""
    )
    escaped = body.replace("<", "&lt;").replace(">", "&gt;")
    return """
    <item>
      <title>%s</title>
      <link>https://kite.kagi.com/%s/1/x</link>
      <description>%s%s%s</description>
      <guid isPermaLink="true">https://kite.kagi.com/%s/1/x</guid>
      <category>%s</category>
      <category>%s/%s</category>
      <category>%s</category>
      <pubDate>%s</pubDate>
    </item>""" % (
        title,
        section.lower(),
        escaped,
        img,
        extras,
        section.lower(),
        section,
        section,
        sub,
        sub,
        pub,
    )


HIGHLIGHTS = (
    "&lt;h3&gt;Highlights:&lt;/h3&gt;&lt;ul&gt;"
    "&lt;li&gt;One thing that happened.&lt;/li&gt;"
    "&lt;li&gt;Another thing that happened.&lt;/li&gt;&lt;/ul&gt;"
)

PERSPECTIVES = (
    "&lt;h3&gt;Perspectives:&lt;/h3&gt;&lt;ul&gt;"
    "&lt;li&gt;A named official: said something. "
    "(&lt;a href='https://outlet.example/a'&gt;Outlet&lt;/a&gt;)&lt;/li&gt;&lt;/ul&gt;"
)

SOURCES = (
    "&lt;h3&gt;Sources:&lt;/h3&gt;&lt;ul&gt;"
    "&lt;li&gt;&lt;a href='https://one.example/a'&gt;Headline one&lt;/a&gt; - one.example&lt;/li&gt;"
    "&lt;li&gt;&lt;a href='https://two.example/b'&gt;Headline two&lt;/a&gt; - two.example&lt;/li&gt;"
    "&lt;li&gt;&lt;a href='https://one.example/a'&gt;Headline one again&lt;/a&gt; - one.example&lt;/li&gt;"
    "&lt;/ul&gt;"
)

FULL_BODY_EXTRAS = HIGHLIGHTS + PERSPECTIVES + SOURCES


# ------------------------------------------------------------------- images


def png(width: int, height: int) -> bytes:
    """A real, decodable single-colour PNG of the given size."""

    def chunk(kind: bytes, payload: bytes) -> bytes:
        return (
            struct.pack(">I", len(payload))
            + kind
            + payload
            + struct.pack(">I", zlib.crc32(kind + payload) & 0xFFFFFFFF)
        )

    ihdr = struct.pack(">IIBBBBB", width, height, 8, 2, 0, 0, 0)
    rows = b"".join(b"\x00" + b"\x7f\x7f\x7f" * width for _ in range(height))
    return (
        b"\x89PNG\r\n\x1a\n"
        + chunk(b"IHDR", ihdr)
        + chunk(b"IDAT", zlib.compress(rows))
        + chunk(b"IEND", b"")
    )


def jpeg(width: int, height: int, *, progressive: bool = False) -> bytes:
    """Enough of a JPEG for a header reader: APP0, then an SOF carrying the size.

    The progressive variant also inserts 0xFF fill bytes before the marker,
    which is the case a scanner that skips only one 0xFF gets wrong.
    """
    sof = b"\xff\xc2" if progressive else b"\xff\xc0"
    fill = b"\xff\xff" if progressive else b""
    app0 = b"\xff\xe0" + struct.pack(">H", 16) + b"JFIF\x00" + b"\x01\x01\x00" + b"\x00\x01" * 2 + b"\x00\x00"
    frame = sof + struct.pack(">HBHHB", 11, 8, height, width, 1) + b"\x01\x11\x00"
    return b"\xff\xd8" + app0 + fill + frame + b"\xff\xda\x00\x08\x01\x01\x00\x00?\x00" + b"\xff\xd9"


def gif(width: int, height: int) -> bytes:
    return b"GIF89a" + struct.pack("<HH", width, height) + b"\x80\x00\x00"


def webp_vp8x(width: int, height: int) -> bytes:
    body = (
        b"VP8X"
        + struct.pack("<I", 10)
        + b"\x00\x00\x00\x00"
        + (width - 1).to_bytes(3, "little")
        + (height - 1).to_bytes(3, "little")
    )
    return b"RIFF" + struct.pack("<I", 4 + len(body)) + b"WEBP" + body


def webp_vp8l(width: int, height: int) -> bytes:
    bits = (width - 1) | ((height - 1) << 14)
    body = b"VP8L" + struct.pack("<I", 5) + b"\x2f" + struct.pack("<I", bits)
    return b"RIFF" + struct.pack("<I", 4 + len(body)) + b"WEBP" + body

File: test/fixtures.ts
/* Synthetic feeds and images.
 *
 * Written by hand rather than captured from Kagi, for two reasons: the repo
 * stays free of third-party article text, and a fixture can carry the cases real
 * feeds happen not to contain today -- a hostile href, a script tag, an item
 * with no picture, the same headline in two sections.
 *
 * The images are headers only. Nothing here decodes a picture, so a valid IHDR
 * and enough padding to clear the byte floor is a complete fixture.
 */

export const BUILT = 'Mon, 10 Aug 2026 12:01:32 +0000';

export function channel(title: string, built: string, items: string): string {
  return `<?xml version='1.0' encoding='UTF-8'?>
<rss xmlns:atom="http://www.w3.org/2005/Atom" version="2.0">
  <channel>
    <title>Kagi News - ${title}</title>
    <link>https://kite.kagi.com/x.xml</link>
    <lastBuildDate>${built}</lastBuildDate>
    ${items}
  </channel>
</rss>`;
}

interface ItemOptions {
  section?: string;
  sub?: string;
  pub?: string;
  body?: string;
  image?: string | null;
  extras?: string;
}

/** One <item>, with its description double-escaped the way a real feed carries
 *  it: the XML holds escaped HTML, which the parser decodes back to markup. */
export function item(title: string, options: ItemOptions = {}): string {
  const {
    section = 'World',
    sub = 'Wildfires',
    pub = 'Sun, 09 Aug 2026 17:13:45 +0000',
    body = '<p>First paragraph.</p><p>Second paragraph.</p>',
    image = 'https://img.example/one.png',
    extras = '',
  } = options;

  const escape = (s: string) => s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
  const picture = image
    ? escape(`<img src='${image}' alt='A caption for the picture.' /><br />`)
    : '';

  return `
    <item>
      <title>${title}</title>
      <link>https://kite.kagi.com/${section.toLowerCase()}/1/x</link>
      <description>${escape(body)}${picture}${extras}</description>
      <guid isPermaLink="true">https://kite.kagi.com/${section.toLowerCase()}/1/x</guid>
      <category>${section}</category>
      <category>${section}/${sub}</category>
      <category>${sub}</category>
      <pubDate>${pub}</pubDate>
    </item>`;
}

const esc = (s: string) => s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');

export const HIGHLIGHTS = esc(
  '<h3>Highlights:</h3><ul><li>One thing that happened.</li><li>Another thing.</li></ul>'
);

export const PERSPECTIVES = esc(
  "<h3>Perspectives:</h3><ul><li>A named official: said something. " +
    "(<a href='https://outlet.example/a'>Outlet</a>)</li></ul>"
);

export const SOURCES = esc(
  '<h3>Sources:</h3><ul>' +
    "<li><a href='https://one.example/a'>Headline one</a> - one.example</li>" +
    "<li><a href='https://two.example/b'>Headline two</a> - two.example</li>" +
    "<li><a href='https://one.example/a'>Headline one again</a> - one.example</li>" +
    '</ul>'
);

export const FULL_EXTRAS = HIGHLIGHTS + PERSPECTIVES + SOURCES;

// --------------------------------------------------------------------- images

/** Pad a header out past `MIN_BYTES`, so the byte floor is never what a test is
 *  accidentally measuring. */
function pad(bytes: number[], to = 100): Uint8Array {
  const out = new Uint8Array(Math.max(bytes.length, to));
  out.set(bytes);
  return out;
}

const be32 = (n: number) => [(n >>> 24) & 255, (n >>> 16) & 255, (n >>> 8) & 255, n & 255];
const be16 = (n: number) => [(n >>> 8) & 255, n & 255];
const le16 = (n: number) => [n & 255, (n >>> 8) & 255];
const le24 = (n: number) => [n & 255, (n >>> 8) & 255, (n >>> 16) & 255];
const chars = (s: string) => [...s].map((c) => c.charCodeAt(0));

export function png(width: number, height: number): Uint8Array {
  return pad([
    0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a,
    ...be32(13), ...chars('IHDR'),
    ...be32(width), ...be32(height),
    8, 2, 0, 0, 0,
    ...be32(0), // CRC, unchecked by the reader
  ]);
}

/** `fill` inserts extra 0xFF bytes before the SOF marker -- the case a scanner
 *  that skips only one 0xFF gets wrong, as progressive JPEGs do in the wild. */
export function jpeg(width: number, height: number, fill = false): Uint8Array {
  return pad([
    0xff, 0xd8,
    ...(fill ? [0xff, 0xff] : []),
    0xff, 0xc0,
    ...be16(11), 8, ...be16(height), ...be16(width), 1, 0x11, 0x00,
    0xff, 0xd9,
  ]);
}

export function gif(width: number, height: number): Uint8Array {
  return pad([...chars('GIF89a'), ...le16(width), ...le16(height), 0x80, 0, 0]);
}

export function webpVP8X(width: number, height: number): Uint8Array {
  return pad([
    ...chars('RIFF'), ...be32(0), ...chars('WEBP'),
    ...chars('VP8X'), ...be32(10), 0, 0, 0, 0,
    ...le24(width - 1), ...le24(height - 1),
  ]);
}

export function webpVP8L(width: number, height: number): Uint8Array {
  const bits = (width - 1) | ((height - 1) << 14);
  return pad([
    ...chars('RIFF'), ...be32(0), ...chars('WEBP'),
    ...chars('VP8L'), ...be32(5), 0x2f,
    bits & 255, (bits >>> 8) & 255, (bits >>> 16) & 255, (bits >>> 24) & 255,
  ]);
}

export function notAnImage(): Uint8Array {
  return pad(chars('<html><body>Forbidden</body></html>'));
}
